deepbox/src/main/java/ch/cyberduck/core/deepbox/DeepboxIdProvider.java

```
@@ -0,0 +1,342 @@
+import ch.cyberduck.core.deepbox.io.swagger.client.model.NodeContent;
+import ch.cyberduck.core.exception.BackgroundException;
+import ch.cyberduck.core.features.FileIdProvider;
+import ch.cyberduck.core.preferences.HostPreferences;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.UUID;
+
+public class DeepboxIdProvider extends CachingFileIdProvider implements FileIdProvider {
+    private static final Logger log = LogManager.getLogger(DeepboxIdProvider.class);
+
+    private final DeepboxSession session;
+    private final int chunksize;
+    private final DeepboxPathContainerService containerService = new DeepboxPathContainerService();
+
+    public DeepboxIdProvider(final DeepboxSession session) {
+        super(session.getCaseSensitivity());
+        this.session = session;
+        this.chunksize = new HostPreferences(session.getHost()).getInteger("deepbox.listing.chunksize");
+    }
+
+    public String getDeepBoxNodeId(final Path file) throws BackgroundException {
+        return this.getFileId(new DeepboxPathContainerService().getDeepboxPath(file));
+    }
+
+    public String getBoxNodeId(final Path file) throws BackgroundException {
+        return this.getFileId(new DeepboxPathContainerService().getBoxPath(file));
+    }
+
+    public String getThirdLevelId(final Path file) throws BackgroundException {
+        return this.getFileId(new DeepboxPathContainerService().getThirdLevelPath(file));
+    }
+
+    private Deque<Path> decompose(final Path path) {
+        final Deque<Path> walk = new ArrayDeque<>();
+        Path next = path;
+        while(!next.isRoot()) {
+            walk.addFirst(next);
+            next = next.getParent();
+        }
+        return walk;
+    }
+
+    @Override
+    public String getFileId(final Path file) throws BackgroundException {
```

Must throw `NotfoundException` instead of `null` return value when not found.